src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1505 @@
+        /// <param name="evalPersmissions">The runtime usage permissions allowed during type inference</param>
+        /// <returns></returns>
+        public static IList<PSTypeName> InferTypeOf(Ast ast, PowerShell powerShell,TypeInferenceRuntimePermissions evalPersmissions)
+        {
+            var context = new TypeInferenceContext(powerShell);
+            using (new ContextPermissionScope(context, evalPersmissions))
+            {
+                return context.InferType(ast, new TypeInferenceVisitor(context)).ToList();
+            }
+        }
+
+        /// <summary>
+        /// Infers the type that the result of executing a statement would have
+        /// </summary>
+        /// <param name="ast">the ast to infer the type from</param>
+        /// <param name="context">The current type inference context</param>
+        /// <param name="evalPersmissions">The runtime usage permissions allowed during type inference</param>
+        /// <returns></returns>
+        internal static IList<PSTypeName> InferTypeOf(Ast ast, TypeInferenceContext context, TypeInferenceRuntimePermissions evalPersmissions = TypeInferenceRuntimePermissions.None)
+        {
+            using (new ContextPermissionScope(context, evalPersmissions)) {
+                return context.InferType(ast, new TypeInferenceVisitor(context)).ToList();
+            }
+        }
+
+        struct ContextPermissionScope : IDisposable
+        {
+            private readonly TypeInferenceRuntimePermissions _permissions;
+            private readonly TypeInferenceContext _context;
+            internal ContextPermissionScope(TypeInferenceContext context, TypeInferenceRuntimePermissions scopePermissions)
+            {
+                _context = context;
+                _permissions = context.RuntimePermissions;
+                _context.RuntimePermissions = scopePermissions;
+            }
+
+            void IDisposable.Dispose()
+            {
+                _context.RuntimePermissions = _permissions;
+            }
+        }
+    }
+
+    internal class TypeInferenceContext
+    {
+        public static readonly PSTypeName[] EmptyPSTypeNameArray = Utils.EmptyArray<PSTypeName>();
+        private readonly PowerShell _powerShell;
+        internal PowerShellExecutionHelper Helper { get; }
+
+        public TypeInferenceContext() : this(PowerShell.Create(RunspaceMode.CurrentRunspace))
+        {
+        }
+
+        public TypeInferenceContext(PowerShell powerShell)
+        {
+            _powerShell = powerShell;
+            Helper = new PowerShellExecutionHelper(powerShell);
+
+        }
+
+        public bool TryGetRepresentativeTypeNameFromExpressionSafeEval(ExpressionAst expression, out PSTypeName typeName)
+        {
+            typeName = null;
+            if (!RuntimePermissions.HasFlag(TypeInferenceRuntimePermissions.AllowSafeEval))
```

Comment:
We should avoid Enum.HasFlag, it's much, much more expensive than a simple bitwise test.

See the implementation here: https://github.com/dotnet/coreclr/blob/e67851210d1c03d730a3bc97a87e8a6713bbf772/src/vm/reflectioninvocation.cpp#L3699